### PR TITLE
feat: Adds `RegionScanner` trait

### PR DESCRIPTION
**Pull request by evenyag**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
- https://github.com/GreptimeTeam/greptimedb/issues/2806
- https://github.com/GreptimeTeam/greptimedb/issues/3886

## What's changed and what's your intention?
This PR adds the `RegionScanner` trait mentioned in https://github.com/GreptimeTeam/greptimedb/issues/3886

```rust
pub trait RegionScanner: Debug + DisplayAs + Send + Sync {
    fn properties(&self) -> &ScannerProperties;
    fn schema(&self) -> SchemaRef;
    fn scan_partition(&self, partition: usize) -> Result<SendableRecordBatchStream, BoxedError>;
}
```

It allows us to split rows to scan into multiple partitions and increase scan parallelism. The `RegionEngine` has a new method `handle_partitioned_query()` to get the `RegionScanner`. It removes the old `handle_query()` method from the trait and only exposes it in the `MitoEngine`.

Since most region engines don't support partitioning, this PR adds a `SinglePartitionScanner` to turn a stream into a region scanner.

The execution plan returned by the dummy table provider now becomes the `ReadFromRegion` plan. This plan adapts a `RegionScanner` to the `ExecutionPlan` trait.

## Checklist

- [ ]  I have written the necessary rustdoc comments.
- [ ]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

File: src/query/src/dummy_catalog.rs [hunk 1]
```diff
@@ -168,12 +168,12 @@ impl TableProvider for DummyTableProvider {
...
 
-        let stream = self
+        let scanner = self
             .engine
-            .handle_query(self.region_id, request)
+            .handle_partitioned_query(self.region_id, request)
             .await
             .map_err(|e| DataFusionError::External(Box::new(e)))?;
-        Ok(Arc::new(StreamScanAdapter::new(stream)))
+        Ok(Arc::new(ReadFromRegion::new(scanner)))
```
Comment (waynexia) [inline src/query/src/dummy_catalog.rs:176]: The struct name changes from a noun to a verb, do want to think it a better name?

File: src/store-api/src/region_engine.rs [hunk 2]
```diff
@@ -172,3 +235,52 @@ pub trait RegionEngine: Send + Sync {
 }
 
 pub type RegionEngineRef = Arc<dyn RegionEngine>;
+
+/// A [RegionScanner] that only scans a single partition.
+pub struct SinglePartitionScanner {
+    stream: std::sync::Mutex<Option<SendableRecordBatchStream>>,
```
Comment (waynexia) [inline src/store-api/src/region_engine.rs:241]:
```suggestion
    stream: Mutex<Option<SendableRecordBatchStream>>,
```

File: src/store-api/src/region_engine.rs [hunk 3]
```diff
@@ -120,6 +124,65 @@ impl From<PbRegionRole> for RegionRole {
...
+    }
+
+    /// Converts to datafusion's [Partitioning].
+    pub fn to_df_partitioning(&self) -> Partitioning {
+        match self {
+            ScannerPartitioning::Unknown(num_partitions) => {
+                Partitioning::UnknownPartitioning(*num_partitions)
+            }
+        }
+    }
```
Comment (waynexia) [inline src/store-api/src/region_engine.rs:149]: We suggest not to provide such a method. Leaving `ScannerPartitioning` internal (in this project) only. DF's `Partitioning` is 
- for generic data partition, especially intermediate data
- hard to add new variants
- not as expressive as we want

And the `ScannerPartitioning` is likely to be used in our optimizer rule, which also bypasses the `Partitioning` interface

File: src/store-api/src/region_engine.rs [hunk 4]
```diff
@@ -120,6 +124,65 @@ impl From<PbRegionRole> for RegionRole {
...
+
+    /// Returns properties of partitions to scan.
+    pub fn partitioning(&self) -> &ScannerPartitioning {
+        &self.partitioning
+    }
+}
+
+/// A scanner that provides a way to scan the region concurrently.
+/// The scanner splits the region into partitions so that each partition can be scanned concurrently.
+pub trait RegionScanner: Debug + DisplayAs + Send + Sync {
```
Comment (waynexia) [inline src/store-api/src/region_engine.rs:173]: Also document this trait is beneath `ExecutionPlan`

File: src/datanode/src/tests.rs [hunk 5]
```diff
@@ -189,11 +188,11 @@ impl RegionEngine for MockRegionEngine {
         Ok(RegionResponse::new(0))
     }
 
-    async fn handle_query(
+    async fn handle_partitioned_query(
         &self,
         _region_id: RegionId,
         _request: ScanRequest,
-    ) -> Result<SendableRecordBatchStream, BoxedError> {
+    ) -> Result<RegionScannerRef, BoxedError> {
```
[the remark this answers is not in the document]
Comment (evenyag) [reply src/datanode/src/tests.rs:195]: Yes, `handle_partitioned_query()` is not a good name. I need a name for the previous `handle_query()` method that returns a stream. Do you have any suggestions?

File: src/store-api/src/region_engine.rs (hunk 3 above)
Comment (evenyag) [reply src/store-api/src/region_engine.rs:149]: We always map our partitions to `unknown` partitioning in DF. The main purpose of this method is to tell df that we can return multiple streams. I can change this to `partition_count()`.

File: src/query/src/dummy_catalog.rs (hunk 1 above)
Comment (evenyag) [reply src/query/src/dummy_catalog.rs:176]: Naming is hard. Do you have suggestions?

File: src/datanode/src/tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (waynexia) [reply src/datanode/src/tests.rs:195]: 👀why not keep it as is, `handle_query()`

File: src/store-api/src/region_engine.rs (hunk 3 above)
Comment (evenyag) [reply src/store-api/src/region_engine.rs]: Removed in https://github.com/GreptimeTeam/greptimedb/pull/3948/commits/537f4b5f47934bf77b47bf499b81100f12f7b42d

Comment (waynexia) [reply src/store-api/src/region_engine.rs]: >We always map our partitions to unknown partitioning in DF.

Makes a little more sense to me. Let's keep it for now and review the usage one month later

Comment (evenyag) [reply src/store-api/src/region_engine.rs]: Already removed. Now only one place uses it.

File: src/datanode/src/tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (evenyag) [reply src/datanode/src/tests.rs:195]: Need a name for `MitoEngine::handle_query() -> Result<SendableRecordBatchStream>` to avoid confusion.

**Review by v0y4g3r — APPROVED** [review]
LGTM

**Review by waynexia — APPROVED** [review]

**Pull request merged**